zebra-network/src/peer/connection.rs

```diff
@@ -368,455 +373,390 @@ where
-                                    _ => unreachable!(
-                                        "unexpected failed connection state while AwaitingResponse: client_receiver: {:?}",
-                                        self.client_rx
-                                    ),
-                                };
-                            }
-                        }
-                        Either::Right((Either::Left(_), _peer_fut)) => {
-                            trace!(parent: &span, "client request timed out");
-                            let e = PeerError::ClientRequestTimeout;
-                            self.state = match self.state {
-                                // Special case: ping timeouts fail the connection.
-                                State::AwaitingResponse {
-                                    handler: Handler::Ping(_),
-                                    ..
-                                } => {
-                                    self.fail_with(e);
-                                    State::Failed
+            }
+            // We're awaiting a response to a client request,
+            // so wait on either a peer message, or on a request cancellation.
+            State::AwaitingResponse {
+                span,
+                mut tx,
+                mut handler,
+                request_timer,
+            } => {
+                // we have to get rid of the span reference so we can tamper with the state
+                let span = span.clone();
+                trace!(parent: &span, "awaiting response to client request");
+                let cancel = future::select(request_timer, tx.cancellation());
+                match future::select(peer_rx.next(), cancel)
+                    .instrument(span.clone())
+                    .await
+                {
+                    Either::Left((None, _)) => Transition::CloseResponse {
+                        e: PeerError::ConnectionClosed.into(),
+                        tx,
+                    },
+                    Either::Left((Some(Err(e)), _)) => {
+                        Transition::CloseResponse { e: e.into(), tx }
+                    }
+                    Either::Left((Some(Ok(peer_msg)), _cancel)) => {
+                        let request_msg = span.in_scope(|| handler.process_message(peer_msg));
+                        // If the message was not consumed, check whether it
+                        // should be handled as a request.
+                        if let Some(msg) = request_msg {
+                            // do NOT instrument with the request span, this is
+                            // independent work
+                            match conn.handle_message_as_request(msg).await {
+                                Ok(()) => {
+                                    Transition::AwaitResponse { tx, handler, span }
+                                    // Transition::AwaitRequest
```

Comment:
Should we be awaiting a request here?

Reply:
I don't think so, I think I initially set it to `AwaitRequest` durring the refactor but that was incorrect and caused panics, so I commented this out when I was fixing it. Just need to go back and delete this commented out code.